### ISSUE TITLE
Obtaining Sermanet's results

### ISSUE THREAD

**Issue by acg32** [acg32/gtsrb#1]
The objective is to get 99.5% on the test set of GTSRB.
- [x] familiarize with Lua
- [x] familiarize with Torch
- [x] set up the data loader
- [x] familiarize with torch.nn
- [x] launch real size experiments
- [x] reproduce the multi-scale architecture
- [x] use the GPU with cunn and cutorch
- [x] fix SGD to really do batch processing
  - [x] reading cifar.torch example 
  - [x] rewrite SGD
- [x] augment the data with jittering
- [x] write a functional test script
  - [x] write a CUDA version
  - [X] solving buffer problem
- [x] preprocess the data
- [ ] obtain 99.5 % on test set
  - [X] update network architecture
  - [x] setup experiments to run sequentially
  - [x] save parameters and link to model files
  - [ ] experiments different hyperparameters
  - [x] check architecture (regarding the results)
    - [x] obtain good results with a 1 layer network https://github.com/antoine-cha/gtsrb/commit/02f21ca94300a66fe4a29e379de7f7412e4bb7eb
    - [x] obtain good results with a 2 layers network https://github.com/antoine-cha/gtsrb/commit/08efde692862f0f76a360922491593b4468b540b
    - [x] fixing my NLL Criterion https://github.com/antoine-cha/gtsrb/commit/ed6c3aaac7a4af1f54563b643d43ea10a658d2aa https://github.com/antoine-cha/gtsrb/commit/24c45ea28cca32991faa677b390659e74849bd61
      - [x] fix the multiscale architecture (90+% training accuracy)  
      - [x] create dataset with translations, scaling and rotations

**Comment by antoine-cha** [comment]
Results after 10 epochs, with Nesterov momentum

| LR | LR Decay | Momentum | Weight Decay | Train Accuracy | Test Accuracy |
| ---- | ------------- | ---------------- | ------------------- | --------------------- | ------------------- |
| 0.1 | 1e-4 | 0.6 | 5e-4 | 46.5 | 8.58 |
| 0.2 | 1e-4 | 0.7 | 5e-4 | 44.09 | 6.33 |
| 0.3 | 1e-4 | 0.4 | 5e-4 | 57.92 | 6.33 |
| 0.05 | 1e-5 | 0.5 | 1e-4 | 36.16 | 4.83 |
| 0.1 | 1e-4 | 0.3 | 5e-4 | 43.27 | 5.67 |

**Comment by antoine-cha** [comment]
After fixing the architecture, I reach 97% train accuracy after 3 epochs with a simpler model.

**Comment by acg32** [comment]
First results, with a 108-108 wide multiscale architecture. 

| Model | MS-1 | MS-2 | MS-3 |
| --- | --- | --- | --- |
| Train accuracy | 46.5 | 44.09 | 57.92 |
| Test accuracy | 8.58 | 6.33 | 6.33 |

The results were obtained thanks to a grid search of hyper parameters.
They are terrible considering the state of the art on this dataset.

**Comment by acg32** [comment]
Results obtained with really thin architectures: 32 units per layer, with different numbers of layers as inidicated.

| Model | 1-layer | 2 layers | Multiscale |
| --- | --- | --- | --- |
| Train accuracy | 90% | 99.1 % | 98.66% |
| Test accuracy | Not available | 1.48 % | 1.05% |

The good results on training were obtained after changing the way I handled the NLL Criterion. https://github.com/antoine-cha/gtsrb/commit/ed6c3aaac7a4af1f54563b643d43ea10a658d2aa https://github.com/antoine-cha/gtsrb/commit/24c45ea28cca32991faa677b390659e74849bd61
I was using a SoftMax on top. It should not be there when using this criterion.  
The architectures can be found there:
- 1-layer network: https://github.com/antoine-cha/gtsrb/commit/02f21ca94300a66fe4a29e379de7f7412e4bb7eb
- 2-layer network: https://github.com/antoine-cha/gtsrb/commit/08efde692862f0f76a360922491593b4468b540b
- Multiscale network: https://github.com/antoine-cha/gtsrb/commit/b107c99060482fb3e50c646d7c45289f793f5119 

The test accuracy is odd, I have to check the testing part.

**Comment by acg32** [comment]
Test results were terrible because I didn't 1-index the classes. https://github.com/antoine-cha/gtsrb/commit/f29a213cec33a9b9b07ecf709d1022d694ea67ec   

| Model | 1-layer | 2 layers | Multiscale |
| --- | --- | --- | --- |
| Test accuracy | Not available | 71.83 % | 80.27% |

Now that the pipeline is functional, I have to widen the multiscale architecture, and use a 2-layer classifier on top.

**Comment by acg32** [comment]
Using a 108-108 wide multiscale architecture with a 2-layer classifier on top, I obtained the following results: 

| Model | Multiscale |
| --- | --- |
| Train accuracy | N.A. (was above 95%) |
| Test accuracy | 93.52% |

Network architecture: https://github.com/antoine-cha/gtsrb/commit/f029b6c348bb930bb37cf77d12c5efd9d5b2ce24
I have to search more hyper parameters now.

**Comment by acg32** [comment]
I led a grid search of hyper parameters with a 108-200 wide architecture. Here are the 2 best on the test set. 

| Configuration | n°1 | n°2 |
| --- | --- | --- |
| Train accuracy | 100.00 % | 99.98 % |
| Test accuracy | **96.38%** | 96.05% |

The parameters were the following:

| Configuration | n°1 | n°2 |
| --- | --- | --- |
| Learning Rate | 1e-2 | 5e-3 |
| LR Decay | 1e-4 | 1e-5 |
| Momentum | 0.6 | 0.4 |
| Weight Decay | 5e-4 | 1e-4 |

The dampening was set to the value of the momentum.

Network architecture: https://github.com/antoine-cha/gtsrb/commit/7159782a6859b09315de9bf203483fe48eedc3e8
The models are overfitting, I have to enhance data augmentation and preprocessing to obtain better performances.

**Comment by acg32** [comment]
I augmented data by creating 3x more images (https://github.com/antoine-cha/gtsrb/commit/3ddcb654ed55b12e02bf7e4101cc81279dd9fccd).

Networks were trained on this augmented dataset.   I am still using the 108-200 wide architecture. 

| Configuration | n°1 | n°2 | n°3 |
| --- | --- | --- | --- |
| Train accuracy | 99.95 % | 99.95 % | 98.96% |
| Test accuracy | **95.43%** | 94.74% | 93.00% |

Compared to past experiences, I greatly increased the weight decay and the LR decay and used Nesterov momentum.
Models were trained during 7 epochs. The batch size was 64.

| Configuration | n°1 | n°2 | n°3 |
| --- | --- | --- | --- |
| Learning Rate | 5e-3 | 5e-3 | 5e-4 |
| LR Decay | 1e-3 | 1e-3 | 1e-4 |
| Momentum | 0.9 | 0.8 | 0.9 |
| Weight Decay | 5e-3 | 1e-3 | 1e-3 |

I will try 108-108 architectures again, also I will thin down the classifier from 100-100 to 100-50. This enables me to increase batch size to 100.
I was using a 3-layer classifier, I moved to 2.

**Comment by acg32** [comment]
My architecture was different from the one in the paper: 
- I was using too much max pooling on the 2-conv branch.
- My classifier was too wide (800 at first layer instead of 100) and too deep (3 layers instead of 2).
- I didn't use max pooling a second time on the 1st stage output that are on the 1-conv branch.

This should prevent overfitting, because the model has been simplified. https://github.com/antoine-cha/gtsrb/commit/16da08a0dc479aca2888462841d9949274626721

**Comment by acg32** [comment]
Here are the new results with the new architecture, I still train on the augmented dataset.

| Configuration | n°1 |
| --- | --- |
| Train accuracy | 99.84 % |
| Test accuracy | **95.38%** |

I used the same parameters as the previous configuration n°1:

| Configuration | n°1 |
| --- | --- |
| Learning Rate | 5e-3 |
| LR Decay | 1e-3 |
| Momentum | 0.9 |
| Weight Decay | 5e-3 |

~~I am going to connect 100 layers to Y channel and 8 to U and V.~~

**Comment by acg32** [comment]
I tested the training on different machines, here are the timing results. 
- Using **cunn** on a 108-108 wide multiscale architecture:
  
  | GPU | Time for 10 batches | Batch size |
  | --- | --- | --- |
  | GTX970 | 1.9s | 64 |
  | GTX750 Ti | 3.4s | 64 |
  | Titan | 1.4s | 64 |
  | Titan | 2.8s | 128 |
  | Titan | 5.5s | 256 |
  | Titan | 10.9s | 512 |

**Comment by acg32** [comment]
- Using **cudnn**  for convolutions and spatial pooling on a 108-108 wide multiscale architecture:
  
  | GPU | Time for 10 batches | Batch size | Speed-up compared to cunn |
  | --- | --- | --- | --- |
  | GTX970 | 1.5s | 64 | 1.27 |
  | GTX750 Ti | 2.8s | 64 | 1.21 |
  | Titan | 1.0s | 64 | 1.4 |
  | Titan | 1.9s | 128 | 1.47 |
  | Titan | 3.8s | 256 | 1.45 |
  | Titan | 7.5s | 512 | 1.45 |
